lib/view_component/base.rb

```diff
@@ -83,6 +83,18 @@ def self._deprecated_generate_mattr_accessor(name)
     _deprecated_generate_mattr_accessor :sidecar
     _deprecated_generate_mattr_accessor :stimulus_controller
 
+    # In order for translations to work in an initializer,
+    # ActionView requires @virtual_path to be set. To set it,
+    # we allocate the component, set @virtual_path, then
+    # call the component's initializer. This is either terrible,
+    # brilliant, or both.
```

Comment:
> This is either terrible, brilliant, or both.

All of the above! 😅 

Another terrible/brilliant proposal is that we only redefine `allocate` and use `super` instead of `send`:

```rb
def self.allocate
  component = super
  component.instance_variable_set(:@virtual_path, virtual_path)
  component
end
```

This should also fix the syntax issues in ruby 2.5 and 2.6, assuming we'll keep supporting them.

---


This doesn't seem to affect performance, so it's just style and communication:

```
Warming up --------------------------------------
             vanilla   541.319k i/100ms
            allocate   589.440k i/100ms
                 new   588.357k i/100ms
Calculating -------------------------------------
             vanilla      5.403M (± 0.7%) i/s -     27.066M in   5.010067s
            allocate      5.891M (± 0.3%) i/s -     29.472M in   5.002889s
                 new      5.890M (± 0.5%) i/s -     30.006M in   5.094568s

Comparison:
            allocate:  5891035.6 i/s
                 new:  5889982.9 i/s - same-ish: difference falls within error
             vanilla:  5402556.1 i/s - 1.09x  (± 0.00) slower

```

```rb
require 'benchmark/ips'

class Vanilla
  def initialize(bar:)
    @foo = foo
    @bar = bar
  end
  
  def foo
    123
  end
end

class Allocate
  def allocate
    object = super
    object.instance_variable_set :@foo, foo
    object
  end
  
  def initialize(bar:)
    @bar = bar
  end
  
  def foo
    123
  end
end

class New
  def new(...)
    object = allocate
    object.instance_variable_set :@foo, foo
    object.initialize(...)
    object
  end
  
  def initialize(bar:)
    @bar = bar
  end
  
  def foo
    123
  end
end

Benchmark.ips do |x|
  x.report("vanilla") { Vanilla.new(bar: 234) }
  x.report("allocate") { Allocate.new(bar: 234) }
  x.report("new") { New.new(bar: 234) }
  x.compare!
end
```